### PR TITLE
Fix for orphaned entity data

### PR DESCRIPTION
**Pull request by ali-graham**
When entity data with the same name but different ids comes in, old keys were being left in the index we keep. This fixes the bug and also has a data migration to clean up any data left in such a state.

@jessereynolds , I'd appreciate it if you could try to think of any dangers I may have missed.

### CODE REVIEW

File: lib/flapjack/data/entity.rb [hunk 1]
```diff
@@ -469,24 +469,34 @@ def self.add(entity, options = {})
...
-          elsif existing_name != entity_name
-            if redis.hexists('all_entity_ids_by_name', entity_name)
-              merge(existing_name, entity_name, :redis => redis)
-            else
-              rename(existing_name, entity_name, :redis => redis) {|multi|
-                multi.hdel('all_entity_ids_by_name', existing_name)
+            unless this_name_original_id.nil?
+              # an entity existed with a matching name but a different id
+              redis.hdel('all_entity_names_by_id', this_name_original_id)
+            end
```
Comment (ali-graham) [inline lib/flapjack/data/entity.rb]: The `unless` clause above is the only functional change in this file, the rest is just cleanup.

File: spec/lib/flapjack/data/entity_spec.rb [hunk 2]
```diff
@@ -490,6 +492,19 @@ def add_state2
...
+
+        Flapjack::Data::Entity.add({'id'   => '5001',
+                                    'name' => 'name2'},
+                                    :redis => @redis)
+
+        expect(@redis.hget('all_entity_ids_by_name', 'name2')).to eq('5001')
+        expect(@redis.hget('all_entity_names_by_id', '5000')).to be_nil
+        expect(@redis.hget('all_entity_names_by_id', '5001')).to eq('name2')
+      end
+
```
Comment (ali-graham) [inline spec/lib/flapjack/data/entity_spec.rb]: NB the old code in `Entity.add` failed against this spec.

File: lib/flapjack/data/migration.rb [hunk 3]
```diff
@@ -125,6 +125,42 @@ def self.migrate_entity_check_data_if_required(options = {})
...
 
+      def self.clear_orphaned_entity_ids(options = {})
+        raise "Redis connection not set" unless redis = options[:redis]
+
+        logger = options[:logger]
+
+        semaphore = obtain_semaphore(ENTITY_DATA_MIGRATION, :redis => redis)
+        if semaphore.nil?
+          unless logger.nil?
+            logger.fatal "Could not obtain lock for data migration (entity id creation). Ensure that " +
```
Comment (jessereynolds) [inline lib/flapjack/data/migration.rb]: Would be good to s/entity id creation/clear orphaned entity ids/ in the logger message

Comment (ali-graham) [inline lib/flapjack/data/migration.rb]: Oops, yup. Will fix now...

File: spec/lib/flapjack/data/entity_spec.rb (hunk 2 above)
Comment (jessereynolds) [inline spec/lib/flapjack/data/entity_spec.rb]: Woo

**Comment by jessereynolds** [comment]
Looks good, merge when ready.

**Pull request merged**